### PR TITLE
tcp: Redefined bind => listen and support for virtual IPs

### PR DESCRIPTION
**Pull request by AndreasAakesson**
`bind()` is now an internal thing for TCP. What was earlier bind is now `listen()`.

It's now possible to set source address or source socket on both `listen()` and `connect()`. These may throw if they are unable to bind to the socket/address. Still missing the wiring with Inet, but added pseudo code: `TCP::validate_address(ip4::Addr)`.

### CODE REVIEW

File: src/net/tcp/tcp.cpp [hunk 1]
```diff
@@ -220,43 +272,72 @@ seq_t TCP::generate_iss() {
...
+void TCP::validate_address(const Address addr)
 {
-  return ((OS::micros_since_boot() >> 10) & 0xffffffff);
-}
-
-void TCP::drop(const tcp::Packet&) {
-  // Stat increment packets dropped
-  packets_dropped_++;
-  debug("<TCP::drop> Packet dropped\n");
+  // TODO:
```
Comment (alfred-bratterud) [inline src/net/tcp/tcp.cpp]: I'll add `inet::is_valid_source(IP4::addr)`

Comment (alfred-bratterud) [inline src/net/tcp/tcp.cpp]: I'll add `Inet::is_valid_source(IP4::addr)`

File: src/net/tcp/tcp.cpp [hunk 2]
```diff
@@ -65,49 +88,76 @@ TCP::TCP(IPStack& inet) :
...
-  }
-  auto& listener = listeners_.emplace(port,
-    std::make_unique<tcp::Listener>(*this, port, std::move(cb))
+  bind(socket);
+
+  auto& listener = listeners_.emplace(socket,
+    std::make_unique<tcp::Listener>(*this, socket, std::move(cb))
     ).first->second;
-  debug("<TCP::bind> Bound to port %i \n", port);
+  debug("<TCP::bind> Bound to socket %s \n", socket.to_string());
```
Comment (alfred-bratterud) [inline src/net/tcp/tcp.cpp]: TCP::listen

File: src/net/tcp/tcp.cpp [hunk 3]
```diff
@@ -65,49 +88,76 @@ TCP::TCP(IPStack& inet) :
...
+Listener& TCP::listen(tcp::Socket socket, ConnectCallback cb)
 {
-  Listeners::const_iterator it = listeners_.find(port);
-  // Already a listening socket.
-  if(it != listeners_.cend()) {
-    throw TCPException{"Port is already taken."};
-  }
-  auto& listener = listeners_.emplace(port,
-    std::make_unique<tcp::Listener>(*this, port, std::move(cb))
+  bind(socket);
```
Comment (alfred-bratterud) [inline src/net/tcp/tcp.cpp]: What will happen on `TCP::listen({{8, 8, 8, 8}, 80}`? I think bind to socket has to verify the source IP with inet.

File: api/net/tcp/tcp.hpp [hunk 4]
```diff
@@ -453,6 +570,83 @@ namespace net {
...
+     */
+    tcp::Socket bind()
+    { return bind(address()); }
+
+    /**
+     * @brief      Validate the address by making sure it's allowed in this context.
+     *
+     * @param[in]  addr  The address
+     */
+    void validate_address(const tcp::Address addr);
```
Comment (alfred-bratterud) [inline api/net/tcp/tcp.hpp]: Maybe refactor to `validate_source(... addr)`?

File: src/net/tcp/tcp.cpp [hunk 5]
```diff
@@ -220,43 +272,72 @@ seq_t TCP::generate_iss() {
...
-  Expensive look up if port is in use.
-*/
-bool TCP::port_in_use(const port_t port) const {
-  if(listeners_.find(port) != listeners_.end())
-    return true;
+void TCP::bind(const Socket socket)
+{
+  if (UNLIKELY( is_bound(socket) ))
+    throw TCP_error{"Socket is already in use: " + socket.to_string()};
+
```
Comment (alfred-bratterud) [inline src/net/tcp/tcp.cpp]: I think you need to call `validate_address` here.

File: src/net/tcp/tcp.cpp (hunk 3 above)
Comment (AndreasAakesson) [inline src/net/tcp/tcp.cpp]: Fixed.

File: src/net/tcp/tcp.cpp [hunk 6]
```diff
@@ -54,6 +53,30 @@ TCP::TCP(IPStack& inet) :
...
+    throw TCP_error{"All ephemeral ports are taken"};
+
+  ephemeral_ = (ephemeral_ == port_ranges::DYNAMIC_END)
+    ? port_ranges::DYNAMIC_START
+    : ephemeral_ + 1;
+
+  // TODO: Avoid wrap around, increment ephemeral to next free port.
+  // while(is_bound(ephemeral_)) ++ephemeral_; // worst case is like 16k iterations :D
+  // need a solution that checks each word of the subset (the dynamic range)
+  Ensures(is_bound(ephemeral_) == false && "Hoped I wouldn't see the day..."); // this may happen...
```
Comment (AndreasAakesson) [inline src/net/tcp/tcp.cpp]: Feel free to dig in xD

**Pull request merged**